cmd/ipfs/daemon.go

```diff
@@ -300,18 +281,146 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
+	if err != nil {
+		return fmt.Errorf("mountHTTPgw: GetConfig() failed: %s", err), nil
+	}
+
+	gatewayMaddr, err := ma.NewMultiaddr(cfg.Addresses.Gateway)
+	if err != nil {
+		return fmt.Errorf("mountHTTPgw: invalid gateway address: %q (err: %s)", cfg.Addresses.Gateway, err), nil
+	}
+
+	writable, writableOptionFound, err := req.Option(writableKwd).Bool()
+	if err != nil {
+		return fmt.Errorf("mountHTTPgw: req.Option(%s) failed: %s", writableKwd, err), nil
+	}
+	if !writableOptionFound {
+		writable = cfg.Gateway.Writable
+	}
+
+	gwLis, err := manet.Listen(gatewayMaddr)
+	if err != nil {
+		return fmt.Errorf("mountHTTPgw: manet.Listen(%s) failed: %s", gatewayMaddr, err), nil
+	}
+	// we might have listened to /tcp/0 - lets see what we are listing on
+	gatewayMaddr = gwLis.Multiaddr()
+
+	if writable {
+		fmt.Printf("Gateway (writable) server listening on %s\n", gatewayMaddr)
+	} else {
+		fmt.Printf("Gateway (readonly) server listening on %s\n", gatewayMaddr)
+	}
+
+	var opts = []corehttp.ServeOption{
+		corehttp.VersionOption(),
+		corehttp.IPNSHostnameOption(),
+		corehttp.GatewayOption(writable),
 	}
+
+	if len(cfg.Gateway.RootRedirect) > 0 {
+		opts = append(opts, corehttp.RedirectOption("", cfg.Gateway.RootRedirect))
+	}
+
+	node, err := req.Context().ConstructNode()
+	if err != nil {
+		return fmt.Errorf("mountHTTPgw: ConstructNode() failed: %s", err), nil
+	}
+
+	errc := make(chan error)
+	go func() {
+		errc <- corehttp.Serve(node, gwLis.NetListener(), opts...)
+	}()
+	return nil, errc
+}
+
+//collects options and opens the fuse mountpoint
+func mountFuse(req cmds.Request) error {
+	cfg, err := req.Context().GetConfig()
+	if err != nil {
+		return fmt.Errorf("mountFuse: GetConfig() failed: %s", err)
+	}
+
+	fsdir, found, err := req.Option(ipfsMountKwd).String()
+	if err != nil {
+		return fmt.Errorf("mountFuse: req.Option(%s) failed: %s", ipfsMountKwd, err)
+	}
+	if !found {
+		fsdir = cfg.Mounts.IPFS
+	}
+
+	nsdir, found, err := req.Option(ipnsMountKwd).String()
+	if err != nil {
+		return fmt.Errorf("mountFuse: req.Option(%s) failed: %s", ipnsMountKwd, err)
+	}
+	if !found {
+		nsdir = cfg.Mounts.IPNS
+	}
+
+	node, err := req.Context().ConstructNode()
+	if err != nil {
+		return fmt.Errorf("mountFuse: ConstructNode() failed: %s", err)
+	}
+
+	err = commands.Mount(node, fsdir, nsdir)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("IPFS mounted at: %s\n", fsdir)
+	fmt.Printf("IPNS mounted at: %s\n", nsdir)
+	return nil
+}
+
+// merge does fan-in of multiple read-only error channels
+// taken from http://blog.golang.org/pipelines
+func merge(cs ...<-chan error) <-chan error {
+	var wg sync.WaitGroup
+	out := make(chan error)
+
+	// Start an output goroutine for each input channel in cs.  output
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan error) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		if c != nil {
```

Comment:
this is incorrect. if any item in `cs` is nil, this will block forever because of `Add(len(cs))`. 

let's stick to the:

```go
{
  wg.Add(1)
  go output(c)
}
```
idiom.